examples/arkode/C_serial/lsrk_analytic.c

```diff
@@ -0,0 +1,307 @@
+  if (check_flag(&flag, "ARKodeSetMaxNumSteps", 1)) { return 1; }
+
+  /* Specify safety factor for user provided DomEig */
+  flag = LSRKStepSetDomEigSafetyFactor(arkode_mem, 1.01);
+  if (check_flag(&flag, "LSRKStepSetDomEigSafetyFactor", 1)) { return 1; }
+
+  /* Specify the Runge--Kutta--Legendre LSRK method */
+  flag = LSRKStepSetMethod(arkode_mem, ARKODE_LSRK_RKL);
+  if (check_flag(&flag, "LSRKStepSetMethod", 1)) { return 1; }
+
+  /* Open output stream for results, output comment line */
+  UFID = fopen("solution.txt", "w");
+  fprintf(UFID, "# t u\n");
+
+  /* output initial condition to disk */
+  fprintf(UFID, " %.16" ESYM " %.16" ESYM "\n", T0, NV_Ith_S(y, 0));
+
+  /* Main time-stepping loop: calls ARKodeEvolve to perform the integration, then
+     prints results.  Stops when the final time has been reached */
+  t    = T0;
+  tout = T0 + dTout;
+  printf("        t           u\n");
+  printf("   ---------------------\n");
+  while (Tf - t > 1.0e-15)
+  {
+    flag = ARKodeEvolve(arkode_mem, tout, y, &t, ARK_NORMAL); /* call integrator */
+    if (check_flag(&flag, "ARKodeEvolve", 1)) { break; }
+    printf("  %10.6" FSYM "  %10.6" FSYM "\n", t,
+           NV_Ith_S(y, 0)); /* access/print solution */
+    fprintf(UFID, " %.16" ESYM " %.16" ESYM "\n", t, NV_Ith_S(y, 0));
+    if (flag >= 0)
+    { /* successful solve: update time */
+      tout += dTout;
+      tout = (tout > Tf) ? Tf : tout;
+    }
+    else
+    { /* unsuccessful solve: break */
+      fprintf(stderr, "Solver failure, stopping integration\n");
+      break;
+    }
+  }
+  printf("   ---------------------\n");
+  fclose(UFID);
+
+  /* Print final statistics */
+  printf("\nFinal Statistics:\n");
+  flag = ARKodePrintAllStats(arkode_mem, stdout, SUN_OUTPUTFORMAT_TABLE);
+
+  /* Print final statistics to a file in CSV format */
+  FID  = fopen("ark_analytic_nonlin_stats.csv", "w");
+  flag = ARKodePrintAllStats(arkode_mem, FID, SUN_OUTPUTFORMAT_CSV);
+  fclose(FID);
+
+  /* check the solution error */
+  flag = check_ans(y, t, reltol, abstol);
+  flag = compute_error(y, t);
+
+  /* Clean up and return */
+  N_VDestroy(y);           /* Free y vector */
+  ARKodeFree(&arkode_mem); /* Free integrator memory */
+  SUNContext_Free(&ctx);   /* Free context */
+
+  return flag;
+}
+
+/*-------------------------------
+ * Functions called by the solver
+ *-------------------------------*/
+
+/* f routine to compute the ODE RHS function f(t,y). */
+static int f(sunrealtype t, N_Vector y, N_Vector ydot, void* user_data)
+{
+  sunrealtype* rdata = (sunrealtype*)user_data; /* cast user_data to sunrealtype */
+  sunrealtype lambda = rdata[0]; /* set shortcut for stiffness parameter */
+  sunrealtype u      = NV_Ith_S(y, 0); /* access current solution value */
+
+  /* fill in the RHS function: "NV_Ith_S" accesses the 0th entry of ydot */
+  NV_Ith_S(ydot, 0) = lambda * u + SUN_RCONST(1.0) / (SUN_RCONST(1.0) + t * t) -
+                      lambda * atan(t);
```

Comment:
This uses double precision math functions. The function should depend on the precision.

Reply:
Trigonometric functions in the example files are always treated similarly. If you think this is an issue, we should address it in the other examples as well. @Steven-Roberts @drreynolds

Reply:
Just include `tgmath.h` (C99 feature) instead of `math.h` and then `atan` will work properly for whichever floating point type is used.

Reply:
`tgmath.h` is not support on Windows with C99, it requires C11

Reply:
In other examples we do something like this at the top of the file
```
#if defined(SUNDIALS_DOUBLE_PRECISION)
#define COS(x)  (cos((x)))
#elif defined(SUNDIALS_SINGLE_PRECISION)
#define COS(x)  (cosf((x)))
#elif defined(SUNDIALS_EXTENDED_PRECISION)
#define COS(x)  (cosl((x)))
#endif
```

Reply:
> `tgmath.h` is not support on Windows with C99, it requires C11

Ah you're right, I missed that.  David's solution is a good one @maggul. There is no need to copy it to examples that are missing it, but please use it in your example (makes it one less we may need to update later).

Reply:
> In other examples we do something like this at the top of the file
> 
> ```
> #if defined(SUNDIALS_DOUBLE_PRECISION)
> #define COS(x)  (cos((x)))
> #elif defined(SUNDIALS_SINGLE_PRECISION)
> #define COS(x)  (cosf((x)))
> #elif defined(SUNDIALS_EXTENDED_PRECISION)
> #define COS(x)  (cosl((x)))
> #endif
> ```

That should work. This is quite common in example files. Do you think I should apply the same modification to all other occurrences?

Reply:
Any new or updated example should use this approach but you don't need to search for occurrences in other examples.